README.md
# Simple Curl Wrapper Component for PHP

This project aims to deliver an easy to use and free as in freedom object oriented php curl command component.

The build status of the current master branch is tracked by Travis CI:
[![Build Status](https://travis-ci.org/bazzline/php_component_curl.png?branch=master)](http://travis-ci.org/bazzline/php_component_curl)
[![Latest stable](https://img.shields.io/packagist/v/net_bazzline/php_component_curl.svg)](https://packagist.org/packages/net_bazzline/php_component_curl)

The scrutinizer status is:
[![code quality](https://scrutinizer-ci.com/g/bazzline/php_component_curl/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/bazzline/php_component_curl/)

The versioneye status is:
[![Dependency Status](https://www.versioneye.com/user/projects/553941560b24225ef6000002/badge.svg?style=flat)](https://www.versioneye.com/user/projects/553941560b24225ef6000002)

Take a look on [openhub.net](https://www.openhub.net/p/php_component_curl).

The current change log can be found [here](https://github.com/bazzline/php_component_curl/blob/master/CHANGELOG.md).

This component is not developed to replace [guzzle](http://docs.guzzlephp.org/en/latest/).

# Example

## By Using The Builder

```php
//it is always good to ship the component with a factory to easy up usage
use Net\Bazzline\Component\Curl\BuilderFactory;
use Net\Bazzline\Component\Curl\Option\Timeout;

$factory    = new BuilderFactory();
$builder    = $factory->create();
$url        = 'http://www.foo.bar';
$timeout    = new Timeout(10);  //set the timeout to 10 seconds

/**
 * you can also use:
 *  //assuming that $dispatcher is an instance of DispatcherInterface
 *  //assuming that $requestFactory is an instance of RequestFactory
 *  $builder->overwriteDispatcher($dispatcher);
 *  $builder->overwriteRequestFactory($requestFactory);
 */

$response = $builder->usePost()
    ->onTheUrl($url)
    ->withTheData($data)
    ->withTheParameters(array('descendingOrderBy' => 'id'))
    ->withTheOption($timeout)
    ->andFetchTheResponse();

/** 
 * you can also use:
 *  $builder->withTheHeaderLine($headLine)    //add the headline you want
 *  $builder->withResponseModifier($modifier) //add the response modifier you want
 */

echo 'content: ' . $response->content() . PHP_EOL;
echo 'content type: ' . $response->contentType() . PHP_EOL;
echo 'error:' . $response->error() . PHP_EOL;
echo 'error code:' . $response->errorCode() . PHP_EOL;
echo 'head lines: ' . var_export($response->headerLines(), true) . PHP_EOL;
echo 'status code: ' . $response->statusCode() . PHP_EOL;
```

## By Using The Request

```php
//it is always good to ship the component with a factory to easy up usage
$factory    = new Net\Bazzline\Component\Curl\RequestFactory();
$request    = $factory->create();
$url        = 'http://www.foo.bar';

$response = $request->get($url);

echo 'content: ' . $response->content() . PHP_EOL;
echo 'content type: ' . $response->contentType() . PHP_EOL;
echo 'error:' . $response->error() . PHP_EOL;
echo 'error code:' . $response->errorCode() . PHP_EOL;
echo 'head lines: ' . var_export($response->headerLines(), true) . PHP_EOL;
echo 'status code: ' . $response->statusCode() . PHP_EOL;
```

## Executable Examples

* [Delete Request](https://github.com/bazzline/php_component_curl/blob/master/example/make_a_delete_request)
* [Get Request](https://github.com/bazzline/php_component_curl/blob/master/example/make_a_delete_request)
* [Patch Request](https://github.com/bazzline/php_component_curl/blob/master/example/make_a_patch_request)
* [Post Request](https://github.com/bazzline/php_component_curl/blob/master/example/make_a_post_request)
* [Put Request](https://github.com/bazzline/php_component_curl/blob/master/example/make_a_put_request)

## More Examples

### Post Request With Basic Authentication

```
//begin of runtime environments
$factory        = new BuilderFactory();
$builder        = $factory->create();
$data           = array(
    'there'     => 'is',
    'no'        => 'foo',
    'without'   => 'a bar'
);
$password       = '<super secret password>';
$username       = 'foo@bar.ru';
$url            = 'https://foo.bar.ru/api/my/rest/endpoint/v1';
//end of runtime environments

//begin building the request
$builder->asJson();
$builder->onTheUrl($url);
$builder->withTheData($data);
$builder->withTheOption(new SetBasicAuthentication());
$builder->withTheOption(new SetUsernameAndPassword($username, $password));
$builder->usePost();
//end building the request

$request = $builder->andFetchTheResponse();
echo PHP_EOL . 'dumping the request' . PHP_EOL;
var_dump($request);
```

# Terms

* [Dispatcher](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/Dispatcher/Dispatcher.php)
    * doing the curl request
    * if you want to use pure curl, use this class
* [Request](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/Request/Request.php)
    * object oriented approach reflecting the request
    * [HeadLine](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/HeadLine/HeadLineInterface.php)
        * object oriented [http headers](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) ([list](https://en.wikipedia.org/wiki/List_of_HTTP_header_fields) of available headers), start a pull request if you need more
    * [Options](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/Option/OptionInterface.php)
        * object oriented curl options, start a pull request if you need more
    * Parameters
        * all the parameters you want to add to your url - they are urlencoded automatically
    * Url
        * the url of your endpoint
* [Response](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/Response/Response.php)
    * object oriented approach reflecting the response
    * [ResponseBehaviour](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/ResponseBehaviour/ResponseBehaviourInterface.php)
        * interface to interact with the response
            * either modify the response (by creating a new one)
            * change the flow by throwing an exception if the response does not fits your needs (as example)
* [Builder](https://github.com/bazzline/php_component_curl/blob/master/source/Net/Bazzline/Component/Curl/Builder/Builder.php)
    * provides a fluent interface to easy up using curl
    * it takes care of all :-)

# Not Available Curl Options

In general, the php version is limiting the available [curl options](http://php.net/manual/en/curl.constants.php).
Furthermore, some options are not implemented because of their hardcoded usage in the *Response* or the *Dispatcher* (you can set it but they would be overwritten).

This options are:

* used in the *Request*
    * CURLOPT_CUSTOMREQUEST
    * CURLOPT_HTTPHEADER
    * CURLOPT_POSTFIELDS
* used in the *Dispatcher*
    * CURLINFO_HEADER_OUT
    * CURLOPT_HEADERFUNCTION
    * CURLOPT_RETURNTRANSFER

If you want to change this, you either have to extend the existing *Request* or *Dispatcher* object.

# Install

## By Hand

```
mkdir -p vendor/net_bazzline/php_component_curl
cd vendor/net_bazzline/php_component_curl
git clone https://github.com/bazzline/php_component_curl .
```

## With [Packagist](https://packagist.org/packages/net_bazzline/php_component_curl)

```
composer require net_bazzline/php_component_curl:dev-master
```

# Links

* http://resttesttest.com/

## Other Components Available

* https://github.com/php-mod/curl
* https://github.com/anlutro/php-curl
* https://github.com/hamstar/curl
* https://github.com/jyggen/curl
* https://github.com/ixudra/Curl
* https://github.com/brodkinca/BCA-PHP-CURL
* https://github.com/miliqi/laravel-curl
* https://github.com/andrefigueira/Lib-Curl

# Final Words

Star it if you like it :-). Add issues if you need it. Pull patches if you enjoy it. Write a blog entry if you use it. [Donate something](https://gratipay.com/~stevleibelt) if you love it :-].


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Open]

### To Add

* to discuss
    * add [dispatcher](https://github.com/jyggen/curl/blob/master/src/Dispatcher.php) or HandlerGenerator/HandlerFactory
        * https://secure.php.net/manual/en/function.curl-init.php
        * https://secure.php.net/manual/en/function.curl-multi-init.php
    * add RequestModifier
        * e.g. for adding the JsonModifier which converts the data into a json, adds the fitting ContentType etc.
    * add support for parallel request execution like in the [stil/curl](https://github.com/stil/curl-easy) library
    * replace current dispatcher and logging dispatcher strategy with an event driven approach (currently only needed for logging)?
* create Request Data Domain Object
* create Request Options Domain Object

### To Change

## [Unreleased]

### Added

* added php 7.2 to the travis test

### Changed

* updated composer.json file and restricted php version to 5.6 as minimum

## [1.0.0](https://github.com/bazzline/php_component_curl/tree/1.0.0) - released at 2017-04-23

### Added

* support for php version 7.1

### Changed

* dropped support for php version below 5.6
* moved release history from the README.md into this dedicated CHANGELOG.md
* replaced "array()" notation with "[]"

## [0.15.1](https://github.com/bazzline/php_component_curl/tree/0.15.1) - released at 2017-03-29

### Changed

* moved phpunit dependency to 5.7

## [0.15.0](https://github.com/bazzline/php_component_curl/tree/0.15.0) - released at 2016-10-27

### Changed

* support to send data when calling http method DELETE

## [0.14.4](https://github.com/bazzline/php_component_curl/tree/0.14.4) - released at 2016-09-20

### Added

* [AcceptLanguate](https://github.com/bazzline/php_component_curl/blob/0.14.4/source/HeaderLine/AcceptLanguage.php) header line

## [0.14.3](https://github.com/bazzline/php_component_curl/tree/0.14.3) - released at 2016-09-19

### Added

* [AcceptEncoding](https://github.com/bazzline/php_component_curl/blob/0.14.3/source/HeaderLine/AcceptEncoding.php) header line
* [Custom](https://github.com/bazzline/php_component_curl/blob/0.14.3/source/HeaderLine/Custom.php) header line
* started more examples section

### Changed

* updated phpunit to 5.5.*

## [0.14.2](https://github.com/bazzline/php_component_curl/tree/0.14.2) - released at 2016-05-30

### Changed

* relaxed dependency to mockery

## [0.14.1](https://github.com/bazzline/php_component_curl/tree/0.14.1) - released at 2016-03-11

### Changed

* removed *.php* for examples and made them executable
* updated dependencies

## [0.14.0](https://github.com/bazzline/php_component_curl/tree/0.14.0) - released at 2016-02-26

### Changed

* removed default header "application/x-www-form-urlencoded; charset=UTF-8" and added new *ContentTypeUtf8Html*

## [0.13.0](https://github.com/bazzline/php_component_curl/tree/0.13.0) - released at 2016-02-17

### Changed

* fixed "ConverteJsonToArayBehaviour" by merging [pull request 3](https://github.com/bazzline/php_component_curl/pull/3)

## [0.12.0](https://github.com/bazzline/php_component_curl/tree/0.12.0) - released at 2016-02-08

### Added

* *curl_close($handler)* in default *Dispatcher*

### Changed

* fixed bug in *BuilderFactory::createRequestFromFactory()*
* fixed bug in *RequestFactory::create()*

## [0.11.0](https://github.com/bazzline/php_component_curl/tree/0.11.0) - released at 2016-02-08

* fixed major bug in *createRequestFromFactory* to *BuilderFactory*
* refactored internals of Builder::andFetchTheResponse()
* refactored internals of Request::execute()
* refactored internals of RequestFactory::create()

## [0.10.0](https://github.com/bazzline/php_component_curl/tree/0.10.0) - released at 08.02.2016

### Added

* public *overwriteRequestFactory()* to *BuilderFactory*

### Changed

* fixed broken links in the readme
* moved to *psr-4* autoloading
* removed dedicated integration test for php 5.3.3

## [0.9.1](https://github.com/bazzline/php_component_curl/tree/0.9.1) - released at 2016-01-14

### Changed

* updated dependencies


source/Request/Request.php
<?php
/**
 * @author: stev leibelt <artodeto@bazzline.net>
 * @since: 2015-12-09
 */

namespace Net\Bazzline\Component\Curl\Request;

use Net\Bazzline\Component\Curl\Dispatcher\DispatcherInterface;
use Net\Bazzline\Component\Curl\HeaderLine\HeaderLineInterface;
use Net\Bazzline\Component\Curl\Option\OptionInterface;
use Net\Bazzline\Component\Curl\Response\Response;
use Net\Bazzline\Component\Toolbox\HashMap\Merge;

class Request
{
    //@see: http://developer.sugarcrm.com/2013/08/30/doing-put-and-delete-with-curl-in-php/
    const HTTP_METHOD_DELETE    = 'DELETE';
    const HTTP_METHOD_GET       = 'GET';
    const HTTP_METHOD_PATCH     = 'PATCH';
    const HTTP_METHOD_POST      = 'POST';
    const HTTP_METHOD_PUT       = 'PUT';

    /** @var array */
    private $defaultHeaderLines = [];

    /** @var array */
    private $defaultOptions = [];

    /** @var DispatcherInterface */
    private $dispatcher;

    /** @var array */
    private $headerLines = [];

    /** @var Merge */
    private $merge;

    /** @var array */
    private $options = [];

    /**
     * @param DispatcherInterface $dispatcher
     * @param Merge $merge
     * @param array $defaultHeaderLines
     *  <int> => <value>
     * @param array $defaultOptions
     *  either:
     *      <int> => <value>
     *  or:
     *      <identifier> => <value>
     */
    public function __construct(DispatcherInterface $dispatcher, Merge $merge, array $defaultHeaderLines = [], array $defaultOptions = [])
    {
        $this->defaultHeaderLines   = $defaultHeaderLines;
        $this->defaultOptions       = $defaultOptions;
        $this->dispatcher           = $dispatcher;
        $this->merge                = $merge;
    }

    /**
     * @return Request
     */
    public function __clone()
    {
        return new self(
            $this->dispatcher,
            $this->merge,
            $this->defaultHeaderLines,
            $this->defaultOptions
        );
    }

    /**
     * @param HeaderLineInterface $line
     */
    public function addHeaderLine(HeaderLineInterface $line)
    {
        $this->headerLines[] = $line->line();
    }

    /**
     * @param OptionInterface $option
     */
    public function addOption(OptionInterface $option)
    {
        $this->options[$option->identifier()] = $option->value();
    }

    /**
     * @param string $line - CURLOPT_* - see: http://php.net/manual/en/function.curl-setopt.php
     */
    public function addRawHeaderLine($line)
    {
        $this->headerLines[] = $line;
    }

    /**
     * @param string $key - CURLOPT_* - see: http://php.net/manual/en/function.curl-setopt.php
     * @param mixed $value
     */
    public function addRawOption($key, $value)
    {
        $this->options[$key] = $value;
    }

    /**
     * @param string $url
     * @param array $parameters
     * @return Response
     */
    public function get($url, array $parameters = [])
    {
        return $this->execute(
            $url,
            self::HTTP_METHOD_GET,
            $parameters,
            []
        );
    }

    /**
     * @param string $url
     * @param array $parameters
     * @param null|string|array $data
     * @return Response
     */
    public function post($url, array $parameters = [], $data = null)
    {
        return $this->execute(
            $url,
            self::HTTP_METHOD_POST,
            $parameters,
            $data
        );
    }

    /**
     * @param string $url
     * @param array $parameters
     * @param null|string|array $data
     * @return Response
     */
    public function put($url, array $parameters = [], $data = null)
    {
        return $this->execute(
            $url,
            self::HTTP_METHOD_PUT,
            $parameters,
            $data
        );
    }

    /**
     * @param string $url
     * @param array $parameters
     * @param null|string|array $data
     * @return Response
     */
    public function patch($url, array $parameters = [], $data = null)
    {
        return $this->execute(
            $url,
            self::HTTP_METHOD_PATCH,
            $parameters,
            $data
        );
    }

    /**
     * @param string $url
     * @param array $parameters
     * @param null|string|array $data
     * @return Response
     */
    public function delete($url, array $parameters = [], $data = null)
    {
        return $this->execute(
            $url,
            self::HTTP_METHOD_DELETE,
            $parameters,
            $data
        );
    }

    /**
     * @see: https://de.wikipedia.org/wiki/Representational_State_Transfer
    public function head($url)
    {
    }
    public function options($url)
    {
    }
    public function trace($url)
    {
    }
     */

    /**
     * @param bool|false $alsoTheDefaults
     */
    public function reset($alsoTheDefaults = false)
    {
        $this->headerLines  = [];
        $this->options      = [];

        if ($alsoTheDefaults) {
            $this->defaultHeaderLines   = [];
            $this->defaultOptions       = [];
        }
    }

    /**
     * @param array $options
     * @param mixed $data
     * @return array
     */
    private function addDataToTheOptionsIfDataIsValid(array $options, $data)
    {
        $isDataProvided = (!is_null($data));

        if ($isDataProvided) {
            $dataIsNotFromTypeScalar   = (!is_scalar($data));

            if ($dataIsNotFromTypeScalar) {
                $data = http_build_query($data);
            }

            $dataStringContainsContent = (strlen($data) > 0);

            if ($dataStringContainsContent) {
                $options[CURLOPT_POSTFIELDS] = $data; //@see: http://www.lornajane.net/posts/2009/putting-data-fields-with-php-curl
            }
        }

        return $options;
    }

    /**
     * @param array $parameters
     * @param string $url
     * @return string
     */
    private function addParametersToTheUrlIfParametersAreProvided(array $parameters, $url)
    {
        $areParametersProvided = (!empty($parameters));

        if ($areParametersProvided) {
            $parametersAsString     = http_build_query($parameters);
            $isParameterStringValid = (strlen($parametersAsString) > 0);

            if ($isParameterStringValid) {
                $urlWithParameters = $url . '?' . http_build_query($parameters);
            } else {
                $urlWithParameters = $url;
            }
        } else {
            $urlWithParameters = $url;
        }

        return $urlWithParameters;
    }

    /**
     * @param string $url
     * @param string $method
     * @param null|array $parameters
     * @param null|string|array $data
     * @return Response
     */
    private function execute($url, $method, array $parameters = [], $data = null)
    {
        //begin of dependencies
        $dispatcher             = $this->dispatcher;
        $merge                  = $this->merge;
        $headerLines            = $merge($this->headerLines, $this->defaultHeaderLines);
        $options                = $merge($this->options, $this->defaultOptions);
        $headerLines[]          = 'X-HTTP-Method-Override: ' . $method; //@see: http://tr.php.net/curl_setopt#109634
        //end of dependencies

        //begin of business logic
        $options[CURLOPT_CUSTOMREQUEST] = $method; //@see: http://tr.php.net/curl_setopt#109634
        $options[CURLOPT_HTTPHEADER]    = $headerLines;
        //@todo what about binary transfer?

        $options            = $this->addDataToTheOptionsIfDataIsValid($options, $data);
        $urlWithParameters  = $this->addParametersToTheUrlIfParametersAreProvided($parameters, $url);
        $response           = $dispatcher->dispatch($urlWithParameters, $options);
        //end of business logic

        return $response;
    }
}
